alsaaudio.c

```diff
@@ -845,13 +864,173 @@ alsapcm_htimestamp(alsapcm_t *self, PyObject *args)
 }
 
 
-PyDoc_STRVAR(pcm_htimestamp_doc,
+PyDoc_STRVAR(htimestamp_doc,
 "htimestamp() -> tuple\n\
 \n\
 Returns a tuple containing the seconds since epoch in the first element \n\
 , nanoseconds in the second element, and number of frames available in \n\
  the buffer at the time of the time stamp. \n");
 
+
+static PyObject *
+alsapcm_set_tstamp_mode(alsapcm_t *self, PyObject *args)
+{
+	snd_pcm_tstamp_t mode = SND_PCM_TSTAMP_ENABLE;
+	int err;
+
+	if (!PyArg_ParseTuple(args,"|i:set_tstamp_mode", &mode))
+		return NULL;
+				 
+	if (!self->handle)
+	{
+		PyErr_SetString(ALSAAudioError, "PCM device is closed");
+		return NULL;
+	}
+
+	snd_pcm_sw_params_t* swParams;
+	snd_pcm_sw_params_alloca( &swParams);
+
+	snd_pcm_sw_params_current(self->handle, swParams);
+
+	snd_pcm_sw_params_set_tstamp_mode(self->handle, swParams, mode);
+
+	err = snd_pcm_sw_params(self->handle, swParams);
+
+	if (err < 0) {
+		PyErr_SetString(PyExc_RuntimeError, "Unable to set pcm tstamp mode!");
+		return NULL;
+	}
+
+	return Py_None;
+}
+
+
+PyDoc_STRVAR(set_tstamp_mode_doc,
+"set_tstamp_mode() -> None\n\
+\n\
+Set the timestamp mode of the device.  \n");
+
+
+static PyObject *
+alsapcm_get_tstamp_mode(alsapcm_t *self, PyObject *args)
+{
+	snd_pcm_tstamp_t mode;
+	int err;
+
+
+	PyObject *value;
+
+	if (!PyArg_ParseTuple(args,":get_tstamp_mode"))
+		return NULL;
+				 
+	if (!self->handle)
+	{
+		PyErr_SetString(ALSAAudioError, "PCM device is closed");
+		return NULL;
+	}
+
+	snd_pcm_sw_params_t* swParams;
+	snd_pcm_sw_params_alloca( &swParams);
+
+	snd_pcm_sw_params_current(self->handle, swParams);
+
+	err = snd_pcm_sw_params_get_tstamp_mode(swParams, &mode);
+
+	if (err < 0) {
+		PyErr_SetString(PyExc_RuntimeError, "Unable to get pcm tstamp mode!");
+		return NULL;
+	}
+
+	value = PyLong_FromUnsignedLong((unsigned long) mode);
+	return value;
+}
+
+
+PyDoc_STRVAR(get_tstamp_mode_doc,
+"get_tstamp_mode() -> integer \n\
+\n\
+Get the timestamp mode of the device.  \n");
+
+
+static PyObject *
+alsapcm_set_tstamp_type(alsapcm_t *self, PyObject *args)
+{
+	snd_pcm_tstamp_type_t type = SND_PCM_TSTAMP_TYPE_GETTIMEOFDAY;
+	int err;
+
+	if (!PyArg_ParseTuple(args,"|i:set_tstamp_type", &type))
+		return NULL;
+				 
+	if (!self->handle)
+	{
+		PyErr_SetString(ALSAAudioError, "PCM device is closed");
+		return NULL;
+	}
+
+	snd_pcm_sw_params_t* swParams;
+	snd_pcm_sw_params_alloca( &swParams);
+
+	snd_pcm_sw_params_current(self->handle, swParams);
+
+	snd_pcm_sw_params_set_tstamp_type(self->handle, swParams, type);
+
+	err = snd_pcm_sw_params(self->handle, swParams);
+
+	if (err < 0) {
+		PyErr_SetString(PyExc_RuntimeError, "Unable to set pcm tstamp type!");
+		return NULL;
+	}
+
+	return Py_None;
```

Comment:
You need to do a `Py_INCREF(Py_None);` here.

Reply:
Done